Remove shell-specific logic and update related command/arg handling (#306)
buildpacks/spec: platform.md

- The lifecycle SHALL execute all buildpacks in the order defined in `<group>` according to the process outlined in the [Buildpack Interface Specification](buildpack.md).
- The lifecycle SHALL add all invoked buildpacks to`<layers>/config/metadata.toml`.
- The lifecycle SHALL aggregate all `processes` and `slices` returned by buildpacks in `<layers>/config/metadata.toml`.

>     - For buildpacks implementing Buildpack API < 0.9, the lifecycle SHALL combine each process' `command` and `args` in `launch.toml` into `command` in `<layers>/config/metadata.toml`, and leave `args` empty in `<layers>/config/metadata.toml`.

Reviewer: ```suggestion
    - For buildpacks implementing a Buildpack API where the `direct` field on process types is supported, the lifecycle SHALL combine each process' `command` and `args` in `launch.toml` into `command` in `<layers>/config/metadata.toml`, and leave `args` empty in `<layers>/config/metadata.toml`.
```

IDK if this is actually better. I just noticed that you used this verbiage below and tried to make it consistent. Using the language "where supported" (vs. specifying the exact Buildpack API version) has given us some much-needed wiggle room in the past.

|------------------------------------|----------------------|---------------|-----------------------------------------------------------|
| `<app>`                            | `CNB_APP_DIR`        | `/workspace`  | Path to application directory                             |
| `<layers>`                         | `CNB_LAYERS_DIR`     | `/layers`     | Path to layer directory                                   |
| `<process-type>`                   |                      |               | `type` of process to launch                               |
| `<cmd>`                            |                      |               | Command to execute                                        |
| `<args>`                           |                      |               | Arguments to command                                      |
| `<layers>/config/metadata.toml`    |                      |               | Build metadata (see [`metadata.toml`](#metadatatoml-toml) |
| `<layers>/<buildpack-id>/<layer>/` |                      |               | Launch Layers                                             |

> A command (`<cmd>`), arguments to that command (`<args>`), and a working directory (`<working-dir>`) comprise a process definition. Processes MAY be buildpack-defined or user-defined.

Reviewer: Do we need to make mention of the `$(env)` interpolation that the launcher provides? The RFC [mentions](https://github.com/buildpacks/rfcs/blob/main/text/0093-remove-shell-processes.md#using-environment-variables-in-a-process)
> If a buildpack or user includes `$(<env>)` in the command or args and `<env>` is the name of an environment variable set in the launch environment, the launcher will replace this string with the value of the environment variable after apply buildpack-provided env modifications and before launching the process.

Reviewer: The unresolved questions [section](https://github.com/buildpacks/rfcs/blob/main/text/0093-remove-shell-processes.md#unresolved-questions) of the RFC asks if we need a mechanism for turning `$(env)` interpolation off... I guess we should resolve that in this PR? 

It would be awesome if others could chime in with real-life examples of how this might be used - the example given `CNB_INTERPOLATE=false` doesn't solve for cases where users might want to skip interpolating __some__ variables but not all. Should that be a requirement? IDK what is the answer here.

Author: Added a description of environment variable interpolation, and went all-in on the [Kubernetes convention](https://kubernetes.io/docs/tasks/inject-data-application/define-interdependent-environment-variables/) (replace if defined, don't replace if undefined, and use `$$(<env>)` to escape). It makes sense to me to follow the k8s convention. Happy to hear any feedback.

| `<layers>/<buildpack-id>/<layer>/` |                      |               | Launch Layers                                             |

A command (`<cmd>`), arguments to that command (`<args>`), a working directory (`<working-dir>`), and an execution strategy (`<direct>`) comprise a process definition. Processes MAY be buildpack-defined or user-defined.
A command (`<cmd>`), arguments to that command (`<args>`), and a working directory (`<working-dir>`) comprise a process definition. Processes MAY be buildpack-defined or user-defined.

The launcher:

> - MUST derive the values of `<cmd>`, `<args>`, `<working-dir>`, and `<direct>` as follows:

Reviewer: Sadly I don't think we can totally do away with the concept of `direct` in the platform spec until buildpack 0.8 has been formally deprecated. This section probably needs to branch on `direct` being true or false.

Also I don't know where exactly this would go, but the RFC points out
> user provided profiles will not be evaluated when using the new platform API, even when a shell process is executed
This seems like a relatively important point for users to understand. We might want to call it out explicitly.

Author: Agreed...this is pretty fundamental...going to mark as draft until we get this ironed out.

Author: Addressed this; ready for another review.

#### `launcher`
Usage:
```
/cnb/process/<process-type> [<args>...]
# OR

> /cnb/lifecycle/launcher [<cmd> <args>...]

Reviewer 2: This is a **very** significant end user change. We should be careful around this since users won't be able to control this. How do you imagine this working in general since processes are global and not buildpack specific? Would processes contributed by the older buildpack apis still allow for the old launcher approach?

| `<process-type>`                   |                      |               | `type` of process to launch                               |
| `<cmd>`                            |                      |               | Command to execute                                        |
| `<args>`                           |                      |               | Arguments to command                                      |
| `<layers>/config/metadata.toml`    |                      |               | Build metadata (see [`metadata.toml`](#metadatatoml-toml) |
| `<layers>/<buildpack-id>/<layer>/` |                      |               | Launch Layers                                             |

A command (`<cmd>`), arguments to that command (`<args>`), and a working directory (`<working-dir>`) comprise a process definition. Processes MAY be buildpack-defined or user-defined.

The launcher:
- MUST derive the values of `<cmd>`, `<args>`, `<working-dir>`, and `<direct>` as follows:
- **If** the final path element in `$0`, matches the type of any buildpack-provided process type
    - `<process-type>` SHALL be the final path element in `$0`

>     - **If** the buildpack API version supports the `direct` flag on process types, **and** the `direct` flag for this process type is set to `false`

Reviewer: ```suggestion
    - **If** the buildpack API version supports the `direct` field on process types, **and** the `direct` flag for this process type is set to `false`
```

        - `<direct>` is `false`
    - **Else**
        - `<direct>` is `true`.
    - The lifecycle:
        - MUST select the process with type equal to `<process-type>` from `<layers>/config/metadata.toml`
        - MUST set `<cmd>` to the first element of `command` in the selected process type.
        - MUST set `<args>` to any remaining elements of `command` in the selected process type, followed by:
            - **If** there are user-provided `<args>`, the user-provided `<args>`
            - **Else** any `args` defined in the selected process type
        - MUST set `<working-dir>` to the `working-dir` defined for the selected process type, or to `<app>` if not defined
- **Else**
    - `<direct>` SHALL be `true`
    - `<cmd>` SHALL be the user-provided `<cmd>`
    - `<args>` SHALL be the user-provided `<args>`
    - `<working-dir>` SHALL be `<app>`

> - **If** `<direct>` is `true`
>     - MUST invoke the command `<cmd>` with its arguments `<args>`, environment `<env>`, and working directory `<working-dir>`.
> - **Else** `<direct>` is `false`, and
>     - MUST invoke the command `<cmd>` using a shell with its arguments `<args>`, environment `<env>`, and working directory `<working-dir>`.

Reviewer: ```suggestion
- MUST invoke the command `<cmd>` with its arguments `<args>`, environment `<env>`, and working directory `<working-dir>`.
```

Reviewer: I think this is right... right? You can't have a direct process if it wasn't defined by a buildpack on an older buildpack api.

Author: It think we need to consider both here...I updated the indentation in 493bb4a to try to show the intention.

Basically, above this part we define how to set `direct`, and here we describe what needs to happen based on the value of `direct`. Does that make sense?

> - MUST replace all occurrences of `$(<env>)` in `<args>`, where `<env>` is the name of a defined environment variable, with the value of the environment variable after it makes buildpack-provided environment modifications, and before it launches the process.

Reviewer: ```suggestion
- MUST replace all occurrences of `$(<env>)` in `<args>`, where `<env>` is the name of a buildpack-defined environment variable, with the buildpack-provided value, before it launches the process.
```

Reviewer: I notice we don't make mention of any mechanism for turning interpolation off completely (https://github.com/buildpacks/rfcs/blob/main/text/0093-remove-shell-processes.md#unresolved-questions) which seems okay to me as I haven't heard anyone specifically ask that we add this - but maybe others have more information than me. cc @ekcasey @samj1912

[[processes]]
type = "<process type>"

> command = ["<command>"]

Reviewer: I think we need to describe (maybe in the builder section) the [conversions](https://github.com/buildpacks/rfcs/blob/main/text/0093-remove-shell-processes.md#api-compat) that are needed from `launch.toml` -> `metadata.toml` to provide a consistent user interface and avoid breaking older buildpacks.

Author: Added the description in the builder section.

}
```
This label MUST contain the JSON representation of [`project-metadata.toml`](#project-metadatatoml-toml)

## Deprecations

### Sourcing `.profile` or `.profile.bat` Scripts

_Deprecated in Platform API 0.10._

The launcher no longer sources `<app>/.profile` or `<app>/.profile.bat` files if present.

> The launcher previously sourced these files when the command was run in a shell (i.e., `direct = false`), before executing the command in the same shell.

Author: Add a note here about the .profile buildpack when that is ready.

}
```
This label MUST contain the JSON representation of [`project-metadata.toml`](#project-metadatatoml-toml)

## Deprecations

### Sourcing `.profile` or `.profile.bat` Scripts

> _Deprecated in Platform API 0.10._

Reviewer: ```suggestion
_Unsupported in Platform API 0.10._
```
This seems more accurate...
